cellxgene_schema_cli/tests/test_schema_compliance.py

```diff
@@ -0,0 +1,1006 @@
+            self.validator.errors,
+            [
+                "ERROR: ' There is a leading space' in 'uns['title']' is not valid, it contains leading spaces."
+            ],
+        )
+
+        self.validator.adata.uns["title"] = "There is a trailing space "
+        self.validator.errors = []
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: 'There is a trailing space ' in 'uns['title']' is not valid, it contains trailing spaces."
+            ],
+        )
+
+        self.validator.adata.uns["title"] = "There are   double   spaces"
+        self.validator.errors = []
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: 'There are   double   spaces' in 'uns['title']' is not valid, it contains double spaces."
+            ],
+        )
+
+    def test_schema_version(self):
+
+        """
+        Schema_version, This MUST be "2.0.0".
+        """
+
+        self.validator.adata.uns["schema_version"] = "1.0.0"
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: Schema version '1.0.0' is not supported. Validation "
+                "cannot be performed."
+            ],
+        )
+
+    def test_title(self):
+
+        """
+        Title MUST be a string
+        """
+
+        # list instead of string
+        self.validator.adata.uns["title"] = ["title"]
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: '['title']' in 'uns['title']' is not valid, "
+                "it must be a string."
+            ],
+        )
+
+    def test_X_normalization_is_str(self):
+
+        """
+        X_normalization str.
+        """
+
+        # list instead of string
+        self.validator.adata.uns["X_normalization"] = ["normalization"]
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: '['normalization']' in 'uns['X_normalization']' is "
+                "not valid, it must be a string."
+            ],
+        )
+
+    def test_X_normalization_not_raw(self):
+
+        """
+        X_normalization str. This SHOULD describe the method used to normalize the data stored in AnnData X.
+        If data in X are raw, this SHOULD be "none".
+
+        FAIL CASE for when X_normalization was set to "none" but X is not raw data
+        """
+
+        # Assign a real value to X while X_normalization is 'none'
+        self.validator.adata.X[0, 0] = 1.5
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: Matrix 'X' seems to be the raw matrix but not all of "
+                "its values are integers."
```

Comment:
In the code the test is actually done using fractional parts, so floats are allowed as long as they do not have decimal values (i.e. referring to mathematical integers)

@brianraymor any comments on whether to include the kind of test that is performed in the actual schema?

Reply:
If our requirements are more restrictive than AnnData, then it would be helpful to define the requirements for the values in the schema.

Reply:
Our schema requirements are in line with the checks, but I agree that it may not be obvious to the user that an "integer" check is being done. I don't think we should include it in the schema as it may add more noise to an already busy section. If this check causes confusion in the future we can add some clarifying text to the schema.